Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm

```diff
@@ -2,75 +2,71 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
+#include "InputCommon/ControllerInterface/OSX/OSXJoystick.h"
+
 #include <algorithm>
 #include <sstream>
 
 #include <Foundation/Foundation.h>
 #include <IOKit/hid/IOHIDLib.h>
 
+#include "Common/Logging/Log.h"
 #include "Common/StringUtil.h"
-#include "InputCommon/ControllerInterface/OSX/OSXJoystick.h"
 
 namespace ciface
 {
 namespace OSX
 {
-Joystick::Joystick(IOHIDDeviceRef device, std::string name)
-    : m_device(device), m_device_name(name), m_ff_device(nullptr)
+void Joystick::AddElements(CFArrayRef elements)
 {
-  // Buttons
-  NSDictionary* buttonDict = @{
-    @kIOHIDElementTypeKey : @(kIOHIDElementTypeInput_Button),
-    @kIOHIDElementUsagePageKey : @(kHIDPage_Button)
-  };
+  for (int i = 0; i < CFArrayGetCount(elements); i++)
+  {
+    IOHIDElementRef e = (IOHIDElementRef)CFArrayGetValueAtIndex(elements, i);
 
-  CFArrayRef buttons =
-      IOHIDDeviceCopyMatchingElements(m_device, (CFDictionaryRef)buttonDict, kIOHIDOptionsTypeNone);
+    uint32_t usage = IOHIDElementGetUsage(e);
+    uint32_t usage_page = IOHIDElementGetUsagePage(e);
+    uint32_t type = IOHIDElementGetType(e);
 
-  if (buttons)
-  {
-    for (int i = 0; i < CFArrayGetCount(buttons); i++)
+    switch (type)
     {
-      IOHIDElementRef e = (IOHIDElementRef)CFArrayGetValueAtIndex(buttons, i);
-      // DeviceElementDebugPrint(e, nullptr);
-
-      AddInput(new Button(e, m_device));
+    case kIOHIDElementTypeCollection:
+      AddElements(IOHIDElementGetChildren(e));
+      continue;
+    case kIOHIDElementTypeOutput:
+      continue;
     }
-    CFRelease(buttons);
-  }
 
-  // Axes
-  NSDictionary* axisDict = @{
-    @kIOHIDElementTypeKey : @(kIOHIDElementTypeInput_Misc),
-    @kIOHIDElementUsagePageKey : @(kHIDPage_GenericDesktop)
-  };
+    IOHIDElementCookie cookie = IOHIDElementGetCookie(e);
 
-  CFArrayRef axes =
-      IOHIDDeviceCopyMatchingElements(m_device, (CFDictionaryRef)axisDict, kIOHIDOptionsTypeNone);
+    // Check for any existing elements with the same cookie
+    auto it = std::find_if(m_cookies.begin(), m_cookies.end(),
+                           [cookie](const auto& ref) { return ref == cookie; });
 
-  if (axes)
-  {
-    std::vector<IOHIDElementRef> elems;
-    for (int i = 0; i < CFArrayGetCount(axes); i++)
-    {
-      IOHIDElementRef e = (IOHIDElementRef)CFArrayGetValueAtIndex(axes, i);
-      // DeviceElementDebugPrint(e, nullptr);
-      uint32_t usage = IOHIDElementGetUsage(e);
+    if (it != m_cookies.end())
+      break;
```

Comment:
Shouldn't this `break` be a `continue` so the rest of the elements are looked at?